src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,14 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
-
-            return IntersectIterator(first, second, null);
+           
+            foreach (TSource element in first)
+            {
+                if (second.Contains(element))
```

Comment:
Also, the impl of `Contains` now hit will always go through `IEqualityComparer<TSource>` which you'd claimed as a problem with the original.